Feature/onboarding

JungMinji08 opened the pull request:
온보딩 페이지 생성했어요.
아까 채팅방에서 말했던 "홈에서 뒤로가기 눌렀을 때 스플래쉬로 가지 않도록 처리하기"랑 마찬가지로 온보딩에서도 뒤로가기 누를 경우 홈이 아닌 앱 종료로 이어지게끔 작업했고(저는 BackPressHandler를 만들었는데 다른 방법 있으면 알려주세여...)

한가지 추가한 로직이 "1.5초 안에 뒤로가기 두 번을 눌러야 앱이 나가진다."인데 온보딩 과정에서의 이탈을 조금이라도 막아보자는 의도로 추가했는데 살짝 앱 전반적으로 들어가야할 부분일 수도 있을거같아서.. 이거 관련해서는 나중에 회의해봐야할 것 같습니다~

JungMinji08 committed:
Add OnboardingScreen and BackPressHandler

- OnboardingScreen 추가하고 네비게이션 연결
- BackPressHandler 추가
    - BackPress 이벤트를 중간에 가로채서 원하는 작업을 수행할 수 있도록 해주는 composable 함수 생성
    - 온보딩 페이지에서는 사용자가 뒤로가기를 누를 경우 바로 앱이 종료되는 것이 아닌 "뒤로가기 버튼을 한번 더 누르면 종료됩니다" 라는 안내문구가 나오고, 1.5초 안에 뒤로가기를 한 번 더 눌러야 앱 종료

Sookhee commented on app/src/main/java/com/mashup/damgledamgle/di/RepositoryModule.kt:
```diff
@@ -0,0 +1,24 @@
+ */
+
+@Module
+@InstallIn(SingletonComponent::class)
```
음 이거 스코프 ActivityComponent로 줄여도 될거같은데 어떻게 생각하시나여??

```kotlin
@Module
@InstallIn(ActivityComponent::class)
```

ddyeon commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/onboarding/OnboardingScreen.kt:
```diff
@@ -0,0 +1,33 @@
+    val context = LocalContext.current
+    var backPressWaitTime by remember { mutableStateOf(0L) }
+
+    BackPressHandler {
+        if (System.currentTimeMillis() - backPressWaitTime >= 1500) { // 1.5초 안에 뒤로가기 두 번 눌러야 앱 종료.
+            backPressWaitTime = System.currentTimeMillis()
+            Toast.makeText(context, "뒤로가기 버튼을 한번 더 누르면 종료됩니다.", Toast.LENGTH_SHORT).show()
+        } else {
+            (context as? Activity)?.finish()
+        }
+    }
```
저는 1.5초 안에 두번 누르면 종료되는 거 좋은 것 같아요!
시간도 적당한 것 같습니다~

ddyeon replied:
OnBoarding이 알림설정 페이지(다음 페이지 임시)로 바뀌는건가용
아니면 1.5초 설정하려고 큰 틀처럼 해놓은건가여

JungMinji08 replied:
네네 마자여 큰 틀이욤! 
OnboardingScreen에서 퍼미션 allow 상태 체크해서 각 페이지로 이동시키려고 생각했는데
일단 해보고 안되면 수정하려고요. ㅎㅎ

ddyeon commented on data/src/main/java/com/mashup/damgledamgle/repository/datastore/DataStoreRepositoryImpl.kt:
```diff
@@ -0,0 +1,80 @@
+ *  Copyright © 2022 MashUp All rights reserved.
+ */
+
+class DataStoreRepositoryImpl @Inject constructor(
```
👍🏻👏🏻

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/onboarding/OnboardingScreen.kt:
```diff
@@ -0,0 +1,33 @@
+    var backPressWaitTime by remember { mutableStateOf(0L) }
+
+    BackPressHandler {
+        if (System.currentTimeMillis() - backPressWaitTime >= 1500) { // 1.5초 안에 뒤로가기 두 번 눌러야 앱 종료.
+            backPressWaitTime = System.currentTimeMillis()
+            Toast.makeText(context, "뒤로가기 버튼을 한번 더 누르면 종료됩니다.", Toast.LENGTH_SHORT).show()
+        } else {
+            (context as? Activity)?.finish()
+        }
```
BackPressHandler는 아마 다른 곳에서도 쓰이게 될 거고 아래 두 번 눌러야 앱 종료 되는 로직도 메인에서 들어갈 가능성이 높겠네요
그럼 이 부분은 재사용성을 위해 따로 빼주는게 좋아보입니다
context만 받는 함수로 빼는 방식도 괜찮을거 같네요

dahyun1226 replied:
+ 이 부분 더 이쁘게 개선해볼수 있을거 같은데
리뷰 이후에도 조금 더 고민해볼게요!

JungMinji08 replied:
좋습니당~~ 일단 수정해봤는데
1) BackPressInterceptor로 네이밍 변경
2) 매개변수로 context: Context, onBackPressed: (()->Unit)? = null을 받음
3) onBackPressed에 함수를 넘기면 그걸 실행하고, 안넘기면 그냥 1.5초 체크해서 액티비티 종료시키는 로직을 실행(기본)

바텀시트 등에서 뒤로가기 눌렀을 때 hidden 처리같은거 시켜야 할 수도 있을거같아서 이렇게 수정했는데 혹시 별로라거나 다른 의견 있으면 주세용~

JungMinji08 replied:
아님 개발 좀 더 진행해보고 다시 수정하는 것도 괜찮을 듯

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/onboarding/OnboardingScreen.kt:
```diff
@@ -0,0 +1,33 @@
+    var backPressWaitTime by remember { mutableStateOf(0L) }
+
+    BackPressHandler {
+        if (System.currentTimeMillis() - backPressWaitTime >= 1500) { // 1.5초 안에 뒤로가기 두 번 눌러야 앱 종료.
```
1500도 상수로 Const같은데 빼줘도 좋을듯해요

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/splash/SplashScreen.kt:
```diff
@@ -46,8 +46,9 @@ fun SplashScreen(navController: NavHostController) {
         } else {
+            navController.navigate("onboarding_screen")
             // TODO(minji): move to onboarding page
-            navController.navigate("home_screen") // 임시
+            // navController.navigate("home_screen") // 임시
```
지워도 될듯하네요

JungMinji08 committed:
Refactor

- 코드리뷰 반영
- BackPressHandler -> BackPressInterceptor로 네이밍 변경
- 뒤로가기 눌렀을 때 기본으로는 1.5초 체크해서 액티비티를 종료시키지만, 바텀시트 등에서 뒤로가기 눌렀을 때는 hidden 등 옵션을 처리할 수 있도록 수정
- SplashScreen.kt에 필요없는 주석 제거

JungMinji08 commented:
코드리뷰 반영 했습니다~
그리고 conflict 해결하느라 리베이스했고, 12일까지 리뷰 더 없으면 머지할게영 👀
